File: Example-Objc/ViewController.m
#import "ViewController.h"
#import <DapiConnect/DapiConnect.h>

@interface ViewController () <UITableViewDataSource, UITableViewDelegate, DPCConnectDelegate>

@property (weak, nonatomic) IBOutlet UITableView *tableView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"AccountCellID"];
    
}


- (IBAction)didTapAddButton:(UIBarButtonItem *)sender {
    [Dapi.sharedInstance presentConnect:self];
}

- (IBAction)didTapTransferButton:(UIBarButtonItem *)sender {
    DPCBankConnection *firstConnection = Dapi.sharedInstance.getConnections.firstObject;
    if (firstConnection) {
        [self payUsing:firstConnection account:nil];
    } else {
        [Dapi.sharedInstance presentConnect:self];
    }
}

// MARK:- UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return Dapi.sharedInstance.getConnections.count;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSArray<DPCAccount *> *accounts = Dapi.sharedInstance.getConnections[section].accounts;
    return accounts.count;
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"AccountCellID" forIndexPath:indexPath];
    DPCAccount *account = Dapi.sharedInstance.getConnections[indexPath.section].accounts[indexPath.row];
    cell.textLabel.numberOfLines = 0;
    cell.textLabel.text = [NSString stringWithFormat:@"IBAN: %@\nNumber: %@", account.iban, account.number];
    return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    DPCBankConnection *connection = [Dapi.sharedInstance getConnections][section];
    return connection.bankName;
}

// MARK:- UITableViewDelegate
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    DPCBankConnection *connection = Dapi.sharedInstance.getConnections[indexPath.section];
    [self payUsing:connection account:connection.accounts[indexPath.row]];
}

- (UISwipeActionsConfiguration *)tableView:(UITableView *)tableView trailingSwipeActionsConfigurationForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    DPCBankConnection *connection = Dapi.sharedInstance.getConnections[indexPath.section];
    
    NSMutableArray<UIContextualAction *> *actions = [NSMutableArray array];
    
    UIContextualAction *identityAction = [UIContextualAction contextualActionWithStyle:(UIContextualActionStyleNormal) title:@"Identity" handler:^(UIContextualAction * _Nonnull action, __kindof UIView * _Nonnull sourceView, void (^ _Nonnull completionHandler)(BOOL)) {
        completionHandler(YES);
        [connection getIdentity:^(DPCIdentity * _Nullable identity, NSError * _Nullable error, NSString * _Nullable operationID) {
            if (error) {
                [self showAlertWithTitle:@"Error" message:error.localizedDescription];
            } else {
                [self showAlertWithTitle:@"DOB" message:identity.dateOfBirth];
            }
        }];
    }];
    
    [actions addObject:identityAction];
    return [UISwipeActionsConfiguration configurationWithActions:actions];
}

// MARK:- DPCConnectDelegate

- (void)connectDidFailConnectingToBankID:(nonnull NSString *)bankID withError:(nonnull NSString *)error {
    NSString *errorTitle = [NSString stringWithFormat:@"%@ Connection Failed", bankID];
    [self showAlertWithTitle:errorTitle message:error];
}

- (void)connectDidSuccessfullyConnectToBankID:(NSString *)bankID connection:(DPCBankConnection *)connection {
    [self.tableView reloadData];
    
    /// Calling APIs from client automatically stores the data for you in `[client.connect getConnections]`
    [connection getAccounts:^(NSArray<DPCAccount *> * _Nullable accounts, NSError * _Nullable error, NSString * _Nullable operationID) {
        [self.tableView reloadData];
    }];
}

- (void)connectDidProceedWithBankID:(nonnull NSString *)bankID userID:(nonnull NSString *)userID {
    // this will be removed in the future
}

- (void)showAlertWithTitle:(NSString *)title message:(NSString *)message {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:(UIAlertActionStyleDefault) handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)payUsing:(DPCBankConnection *)connection account:(DPCAccount *)account {
    DPCBeneficiary *beneficiary = [[DPCBeneficiary alloc] init];
    DPCLinesAddress *lineAddress = [[DPCLinesAddress alloc] init];
    lineAddress.line1 = @"baniyas road";
    lineAddress.line2 = @"dubai";
    lineAddress.line3 = @"united arab emirates";
    beneficiary.linesAddress =  lineAddress;
    beneficiary.accountNumber = @"123456789";
    beneficiary.bankName = @"Emirates NBD Bank PJSC";
    beneficiary.swiftCode = @"EBILAEAD";
    beneficiary.iban = @"AE123456789";
    beneficiary.country = @"UNITED ARAB EMIRATES";
    beneficiary.branchAddress = @"Baniyas Road Deira PO Box 777 Dubai UAE";
    beneficiary.branchName = @"Emirates NBD Bank PJSC";
    beneficiary.phoneNumber = @"0123456789";
    beneficiary.name = @"John Doe";
    
    [connection createTransferFromAccount:account toBeneficiary:beneficiary completion:^(DPCAccount * _Nullable account, NSUInteger amount, NSError * _Nullable error, NSString * _Nullable operationID) {
        if (error) {
            [self showAlertWithTitle:@"Error" message:error.localizedDescription];
        } else {
            [self showAlertWithTitle:@"Transfer success" message:[NSString stringWithFormat:@"Transfer from account: %@", account.accountID]];
        }
    }];
}

@end

File: Example-Swift/ViewController.swift
import UIKit
import DapiConnect

class ViewController: UIViewController {
    
    @IBOutlet weak var tableView: UITableView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "AccountCellID")
    }
    
    @IBAction func didTapAddButton(_ sender: UIBarButtonItem) {
        Dapi.shared.presentConnect(self)
    }
    
    @IBAction func didTapTransferButton(_ sender: UIBarButtonItem) {
        if let connection = Dapi.shared.getConnections().first {
            self.pay(amount: 0, using: connection, account: nil)
        }
    }
}

extension ViewController: UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        let connections = Dapi.shared.getConnections()
        return connections.count
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let connectionAccounts = Dapi.shared.getConnections()[section].accounts
        return connectionAccounts.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "AccountCellID", for: indexPath)
        let account = Dapi.shared.getConnections()[indexPath.section].accounts[indexPath.row]
        cell.textLabel?.numberOfLines = 0
        cell.textLabel?.text = "IBAN: \(account.iban)" + "\n" + "Number: \(account.number)"
        return cell
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        let connection = Dapi.shared.getConnections()[section]
        return connection.bankName
    }
}

extension ViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath) -> UISwipeActionsConfiguration? {
        let connection = Dapi.shared.getConnections()[indexPath.section]
        let account = connection.accounts[indexPath.row]
        
        let getTransactionsAction = UIContextualAction.init(style: .normal, title: "Transactions", handler: { (action, view, isPerformed) in
            isPerformed(true)
            let twoDaysAgo = Date().addingTimeInterval(-172800)
            let yesterday = Date().addingTimeInterval(-86400)

            connection.getTransactionsFor(account, from: twoDaysAgo, to: yesterday) { (transactions, error, operationID) in
                switch transactions {
                case .some(let transactions):
                    transactions.forEach { print($0.transactionDescription) }
                    self.showAlert(title: "Transactions", message: "Found \(transactions.count) transaction(s)")
                case .none:
                    self.showAlert(title: "Error", message: error?.localizedDescription ?? "Error fetching balance")
                }
            }
        })
        
        let delinkConnection = UIContextualAction.init(style: .destructive, title: "Delink Connection", handler: { (action, view, isPerformed) in
            isPerformed(true)
            connection.delete { (result, error) in
                if let _ = result {
                    self.tableView.reloadData()
                } else {
                    self.showAlert(title: "Error", message: error?.localizedDescription ?? "Error fetching balance")
                }
            }
        })
        
        return UISwipeActionsConfiguration(actions: [getTransactionsAction, delinkConnection])
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let connection = Dapi.shared.getConnections()[indexPath.section]
        self.pay(amount: 0, using: connection, account: connection.accounts[indexPath.row])
    }
}

extension ViewController: DapiConnectDelegate {
    func connectDidSuccessfullyConnect(toBankID bankID: String, connection: DapiBankConnection) {
        tableView.reloadData()
        /// Calling APIs from client automatically stores the data for you in `client.connect.getConnections()`
        connection.getAccounts { (accounts, error, jobID) in
            self.tableView.reloadData()
        }
    }
    
    func connectDidDismiss() {
        print("connectDidDismiss")
    }
    
    func connectDidFailConnecting(toBankID bankID: String, withError error: String) {
        let errorTitle = "\(bankID) Connection Failed"
        showAlert(title: errorTitle, message: error)
    }
    
    func connectDidProceed(withBankID bankID: String, userID: String) {
        // this will be removed in the future
    }
}

extension ViewController {
    func showAlert(title: String, message: String) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Ok", style: .default))
        present(alert, animated: true)
    }
    
    func pay(amount: UInt, using connection: DapiBankConnection, account: DapiAccount?) {
        let beneficiary = DapiBeneficiary()
        let lineAddress = DapiLinesAddress()
        lineAddress.line1 = "baniyas road"
        lineAddress.line2 = "dubai"
        lineAddress.line3 = "united arab emirates"
        beneficiary.linesAddress =  lineAddress
        beneficiary.accountNumber = "123456789"
        beneficiary.bankName = "Emirates NBD Bank PJSC"
        beneficiary.swiftCode = "EBILAEAD"
        beneficiary.iban = "AE123456789"
        beneficiary.country = "UNITED ARAB EMIRATES"
        beneficiary.branchAddress = "Baniyas Road Deira PO Box 777 Dubai UAE"
        beneficiary.branchName = "Emirates NBD Bank PJSC"
        beneficiary.phoneNumber = "0123456789"
        beneficiary.name = "John Doe"
        connection.createTransfer(from: account,
                                  to: beneficiary,
                                  amount: 1,
                                  completionWithReferenceNumber: { (account, amount, error, id, referenceNumber) in
            print(referenceNumber, error)
        }, remark: "remark")
        
    }
}
